### PR TITLE
Add PENDING status to IBC / ICA function calls

### PR DESCRIPTION
**Pull request by asalzmann**
## Context and purpose of the change

Currently, our ICA / IBC logic assumes efficient relayers. In practice, this has not been the case, and packets sometimes fail to relay. To improve the stability of our ICA / IBC logic and make it easier to reason about the current state, this PR adds a PENDING flag to all ICA / IBC function calls. Specifically, the PENDING status is tracked on various records.

The following functions are async and need to be accounted for in a PENDING state:

**Staking direction** (represented on `DepositRecord`s)
- IBC transfer from Stride --> Host Zone
- ICA bank send from withdrawal account -> delegation account (NOTE: we don't need a PENDING state here, since the record is created in the callback upon a successful MsgSend)
- ICA delegation call

**Unstaking direction** (represented on `EpochUnbondingRecords`s)
- ICA unstaking call
- ICA bank send from the delegation ICA to the redemption ICA

Claims already have a pending state (to prevent double claims).

The logic is as follows: after the ICA / IBC message is triggered, the state of the record is set to `PENDING`. Then, in the callback, the state either (1) progresses based on the action taken, if the ICA / IBC call was a success, or (2) reverts to the previous state, if the call failed.

Took the opportunity here to rename some of the flags to make them more clear.
Also note, the updates to PENDING status should happen as close to the ICA call site as possible. Please check this as you review!

## Brief Changelog

Hooks changes
- Add `PENDING` flag to `DepositRecord`
`TRANSFER` -> `TRANSFER_QUEUE`
`STAKE` -> `DELEGATION_QUEUE`
```
enum Status {
    // in transfer queue to be sent to the delegation ICA
    TRANSFER_QUEUE = 0;
    // transfer in progress (IBC packet sent, ack not received)
    TRANSFER_IN_PROGRESS = 2;
    // in staking queue on delegation ICA
    DELEGATION_QUEUE = 1;
    // staking in progress (ICA packet sent, ack not received)
    DELEGATION_IN_PROGRESS = 3;
  }
```
- Add `PENDING` flag to `EpochUnbondingRecord`
`BONDED`  -> `UNBONDING_QUEUE`
`UNBONDED` -> `EXIT_TRANSFER_QUEUE`
`TRANSFERRED` -> `CLAIMABLE`
```
enum Status {
    // tokens bonded on delegate account
    UNBONDING_QUEUE = 0;
    UNBONDING_IN_PROGRESS = 3;
    // unbonding completed on delegate account
    EXIT_TRANSFER_QUEUE = 1;
    EXIT_TRANSFER_IN_PROGRESS = 4;
    // transfer success
    CLAIMABLE = 2;
  }
```
- In `TransferExistingDepositsToHostZones`, after IBC transferring tokens to the host zone
- In `StakeExistingDepositsOnHostZones`, set the deposit record to status `PENDING` after sending a delegation ICA
- Set `HostZoneUnbonding`s to `PENDING` after transferring tokens to the redemption account in `SweepAllUnbondedTokens`
- Set `HostZoneUnbonding`s to `PENDING` after sending unstaking ICAs

Callback  changes
- Update `DepositRecord` status in `DelegateCallback`
- Update `DepositRecord` status in `TransferCallback`
- Update `HostZoneUnbonding` in `UndelegateCallback`
- Update `HostZoneUnbonding` in `RedemptionCallback`

## Tests
Force (1) timeout and (2) transaction failure for 
- IBC transfer
1. DepostRecord has status TRANSFER_QUEUE
2. DepostRecord has status TRANSFER_IN_PROGRESS
3. DepostRecord has status TRANSFER_QUEUE
timeout: set IBC transfer timeout very low
```
// set timeout to 1s
// observe record change from
TRANSFER_QUEUE -> TRANSFER_IN_PROGRESS -> TRANSFER_QUEUE
```
failure: transfer to an invalid account
```
// set delegateAddress to osmo1j6vwz24hddx4t6kukmzgmk0764x0evs74002e3
// observe record change from
TRANSFER_QUEUE -> TRANSFER_IN_PROGRESS -> TRANSFER_QUEUE
```

- ICA MsgDelegate
1. DepostRecord has status DELEGATION_QUEUE
2. DepostRecord has status DELEGATION_IN_PROGRESS
3. DepostRecord has status DELEGATION_QUEUE
timeout: set ICA timeout very low
```
if callbackId == "delegate" {
	timeoutTimestamp = uint64(ctx.BlockTime().UnixNano()) + uint64(1000000000)
}
// observe the following state transition
DELEGATION_QUEUE -> DELEGATION_IN_PROGRESS -> DELEGATION_QUEUE 
```
failure: stake to non-existent validator 
```
// stake to stridevaloper1dqpg0hfva0k0awzhxwarjpgzj82hlmvrnxlzsd
// observe the following state transition
DELEGATION_QUEUE -> DELEGATION_IN_PROGRESS -> DELEGATION_QUEUE
```

- ICA MsgUndelegate
1. HZU has status UNBONDING_QUEUE
2. HZU has status UNBONDING_IN_PROGRESS
3. HZU has status UNBONDING_QUEUE
timeout: set ICA timeout very low
```
if callbackId == "undelegate" {
	timeoutTimestamp = uint64(ctx.BlockTime().UnixNano()) + uint64(1000000000)
}
// observe the following state transition

```
failure: unstake from non-existent validator 
```
// unstake from stridevaloper1dqpg0hfva0k0awzhxwarjpgzj82hlmvrnxlzsd
// observe the following state transition
UNBONDING_QUEUE -> UNBONDING_IN_PROGRESS -> UNBONDING_QUEUE
```

- ICA MsgSend
1. HZU has status EXIT_TRANSFER_QUEUE
2. HZU has status EXIT_TRANSFER_IN_PROGRESS
3. HZU has status EXIT_TRANSFER_QUEUE
timeout: set ICA timeout very low
```
// set timeout to 1s 
// observe the following state transition
EXIT_TRANSFER_QUEUE -> EXIT_TRANSFER_IN_PROGRESS -> EXIT_TRANSFER_QUEUE
```
failure: send to invalid address
```
// MsgSend to the wrong address
// observe the following state transition
EXIT_TRANSFER_QUEUE -> EXIT_TRANSFER_IN_PROGRESS -> EXIT_TRANSFER_QUEUE
```

## Author's Checklist

I have...

- [ ] Run and PASSED locally all GAIA integration tests
- [ ] If the change is contentful, I either:
    - [ ] Added a new unit test OR 
    - [ ] Added test cases to existing unit tests
- [ ] OR this change is a trivial rework / code cleanup without any test coverage

If skipped any of the tests above, explain.

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] manually tested (if applicable)
- [ ] confirmed the author wrote unit tests for new logic
- [ ] reviewed documentation exists and is accurate

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes? 
  - [ ] Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?
  - [ ] This pull request updates existing proto field values (and require a backend and frontend migration)? 
  - [ ] Does this pull request change existing proto field names (and require a frontend migration)?
  How is the feature or change documented? 
      - [ ] not applicable
      - [ ] jira ticket `XXX` 
      - [ ] specification (`x/<module>/spec/`) 
      - [ ] README.md 
      - [ ] not documented

### CODE REVIEW

File: proto/records/genesis.proto [hunk 1]
```diff
@@ -59,6 +59,8 @@ message DepositRecord {
     TRANSFER = 0;
     // pending staking on delegate account
     STAKE = 1;
+    // ICA call is pending
```
Comment (sampocs) [inline proto/records/genesis.proto:62]: it would be nice to differentiate whether it's pending from an IBC transfer or delegation ICA. Maybe PENDING_TRANSFER and PENDING_STAKE?

This also might not be a bad time to re-evaluate these status fields in general. TRANSFER and STAKE are a little unintuitive.

File: x/records/module_ibc.go [hunk 2]
```diff
@@ -179,17 +179,18 @@ func (im IBCModule) OnAcknowledgementPacket(
...
+	err := im.keeper.ICACallbacksKeeper.CallRegisteredICACallback(ctx, packet, &ack)
+	if err != nil {
+		errMsg := fmt.Sprintf("Unable to call registered callback from records OnAcknowledgePacket | Sequence %d, from %s %s, to %s %s | Error %s",
+			packet.Sequence, packet.SourceChannel, packet.SourcePort, packet.DestinationChannel, packet.DestinationPort, err.Error())
+		im.keeper.Logger(ctx).Error(errMsg)
+		return sdkerrors.Wrapf(icacallbacktypes.ErrCallbackFailed, errMsg)
+	}
+
+	// log the ack type
 	switch resp := ack.Response.(type) {
```
Comment (sampocs) [inline x/records/module_ibc.go:191]: since all this does is log the ack type, might make sense to move this above `CallRegisteredICACallback`

**Review by sampocs — COMMENTED** [review]
Thanks for doing this - this is a game changer!

Couple meta comments:
* Were you able to test this locally? We should confirm the timeout and failed transaction branches in the callbacks. You can force each by modifying the timeout arg and message respectively (e.g. delegate to a validator that doesn't exist). It'd also be nice to confirm that tokens are refunded in these scenarios.
* Might be good to re-think the deposit record statuses. I think at the least it might make sense to differentiate the two PENDINGS (transfer vs stake). But it's also probably a good opportunity to name each status a little more clearly. 
  * We could probably do better than this but first thing that came to mind was: PENDING_TRANSFER, TRANSFERRING, PENDING_STAKE, STAKING?

File: proto/records/genesis.proto (hunk 1 above)
Comment (asalzmann) [reply proto/records/genesis.proto]: Renamed

File: x/stakeibc/keeper/unbonding_records.go [hunk 3]
```diff
@@ -46,7 +46,7 @@ func (k Keeper) CreateEpochUnbondingRecord(ctx sdk.Context, epochNumber uint64)
 // - total amount to unbond
 // - marshalled callback args
 // - error
-func (k Keeper) GetHostZoneUnbondingMsgs(ctx sdk.Context, hostZone types.HostZone) ([]sdk.Msg, uint64, []byte, error) {
+func (k Keeper) GetHostZoneUnbondingMsgs(ctx sdk.Context, hostZone types.HostZone) ([]sdk.Msg, uint64, []byte, []uint64, error) {
```
[the remark this answers is not in the document]
Comment (asalzmann) [reply x/stakeibc/keeper/unbonding_records.go:50]: added

File: x/records/module_ibc.go (hunk 2 above)
Comment (asalzmann) [reply x/records/module_ibc.go:191]: done

File: x/records/keeper/callback_transfer.go [hunk 4]
```diff
@@ -33,12 +33,36 @@ func (k Keeper) UnmarshalTransferCallbackArgs(ctx sdk.Context, delegateCallback
...
+	depositRecord, found := k.GetDepositRecord(ctx, transferCallbackData.DepositRecordId)
+	if !found {
+		k.Logger(ctx).Error(fmt.Sprintf("TransferCallback deposit record not found, packet %v", packet))
+		return sdkerrors.Wrapf(types.ErrUnknownDepositRecord, "deposit record not found %d", transferCallbackData.DepositRecordId)
+	}
+
+	// which method to use for error handling?
+	// if ack.Response.(type) == *channeltypes.Acknowledgement_Error {
+	// 	// handle error
+	// }
```
[the remark this answers is not in the document]
Comment (asalzmann) [reply x/records/keeper/callback_transfer.go:52]: done

**Comment by asalzmann** [comment]
@sampocs this should be ready for a re-review. I cleaned up some of the names and addressed your comments.

I'm running the integration tests now and constructing the failure cases you outlined

File: scripts/start_network.sh [hunk 5]
```diff
@@ -12,7 +12,8 @@ mkdir -p $SCRIPT_DIR/logs
 HERMES_LOGS=$SCRIPT_DIR/logs/hermes.log
 ICQ_LOGS=$SCRIPT_DIR/logs/icq.log
 
-HOST_CHAINS=(GAIA JUNO OSMO STARS)
+# TODO: revert
+HOST_CHAINS=(GAIA) # JUNO OSMO STARS)
```
Comment (sampocs) [inline scripts/start_network.sh:16]: bump on the todo

File: utils/utils.go [hunk 6]
```diff
@@ -16,7 +16,7 @@ import (
 )
 
 var ADMINS = map[string]bool{
-	"stride1k8c2m5cn322akk5wy8lpt87dd2f4yh9azg7jlh": true, // F5
+	"stride1u20df3trc2c2zdhm8qvh2hdjx9ewh00sv6eyy8": true, // F5
```
Comment (sampocs) [inline utils/utils.go:19]: revert

File: x/epochs/README.md [hunk 7]
```diff
@@ -131,26 +129,20 @@ type Keeper interface {
...
-5. creates and deposits records on each host zone
-6. sets withdrawal addresses
-7. updates redemption rates (if the epoch coincides with the correct interval)
-8. processes `TRANSFER` deposit records to the delegation Interchain Account (if the epoch coincides with the correct interval)
-9. processes `STAKE` deposit records to the delegation Interchain Account (if the epoch coincides with the correct interval)
-10. Query the rewards account using interchain queries, with the transfer callback to a delegation account as a staked record (if at proper interval)
+If in a `stride_epoch` identifier it: 5. creates and deposits records on each host zone 6. sets withdrawal addresses 7. updates redemption rates (if the epoch coincides with the correct interval) 8. processes `TRANSFER_QUEUE` deposit records to the delegation Interchain Account (if the epoch coincides with the correct interval) 9. processes `DELEGATION_QUEUE` deposit records to the delegation Interchain Account (if the epoch coincides with the correct interval) 10. Query the rewards account using interchain queries, with the transfer callback to a delegation account as a staked record (if at proper interval)
```
Comment (sampocs) [inline x/epochs/README.md:138]: great call updating this!!

Comment (sampocs) [inline x/epochs/README.md:138]: I know it doesn't affect the actual rendering but it's a little easier to consume the markdown if the newlines are still there between each number

File: x/records/keeper/callback_transfer.go [hunk 8]
```diff
@@ -53,15 +53,15 @@ func TransferCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack
 	if ack.GetError() != "" {
 		// error on host chain
 		// reset the deposit record to status TRANSFER
```
Comment (sampocs) [inline x/records/keeper/callback_transfer.go]: nit: comment, TRANSFER →TRANSFER_QUEUE

File: x/epochs/README.md (hunk 7 above)
Comment (asalzmann) [reply x/epochs/README.md:138]: hm odd I think this was a linter thing

File: x/stakeibc/keeper/deposit_records.go [hunk 9]
```diff
@@ -47,8 +47,8 @@ func (k Keeper) TransferExistingDepositsToHostZones(ctx sdk.Context, epochNumber
 		pstr := fmt.Sprintf("\t[TransferExistingDepositsToHostZones] Processing deposits {%d} {%s} {%d}", depositRecord.Id, depositRecord.Denom, depositRecord.Amount)
 		k.Logger(ctx).Info(pstr)
 
-		// if a TRANSFER_QUEUE record has 0 balance and was created in the previous epoch, it's safe to remove since it will never be updated or used"
-		if depositRecord.Amount <= 0 {
+		// if a TRANSFER_QUEUE record has 0 balance and was created in the previous epoch, it's safe to remove since it will never be updated or used
+		if depositRecord.Amount <= 0 && depositRecord.DepositEpochNumber < epochNumber {
```
Comment (sampocs) [inline x/stakeibc/keeper/deposit_records.go:51]: I think you can remove this extra condition - it's accounted for [here](https://github.com/Stride-Labs/stride/pull/296/commits/eff92d824f21ff19e54f81d03481f5d2b9075169#diff-8edb9f84b67f6d1f48651b601161499f94253b8fce28ce0a405b9d2f50e19dd6R39)

File: x/records/keeper/callback_transfer.go [hunk 10]
```diff
@@ -46,18 +46,23 @@ func TransferCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack
...
+	case *channeltypes.Acknowledgement_Error:
+		k.Logger(ctx).Error(fmt.Sprintf("TransferCallback error %v", response.Error))
 		// error on host chain
 		// put record back in the TRANSFER_QUEUE
 		depositRecord.Status = types.DepositRecord_TRANSFER_QUEUE
 		k.SetDepositRecord(ctx, depositRecord)
 		k.Logger(ctx).Error(fmt.Sprintf("Error  %s", ack.GetError()))
 		return nil
+	default:
+		k.Logger(ctx).Error(fmt.Sprintf("TransferCallback unknown ack response type %v", response))
```
Comment (sampocs) [inline x/records/keeper/callback_transfer.go:72]: Can we just do something like:

```go

if ack.Response.(type) == *channeltypes.Acknowledgement_Error {
    ... handle failure stuff
}
```
Since we know it **has** to be one of Result and Error, and we don't want to do anything for the Result case

File: x/stakeibc/keeper/unbonding_records.go [hunk 11]
```diff
@@ -231,24 +231,7 @@ func (k Keeper) InitiateAllHostZoneUnbondings(ctx sdk.Context, dayNumber uint64)
...
-				hostZoneUnbonding.Status = recordstypes.HostZoneUnbonding_UNBONDING_IN_PROGRESS
-				// save the updated hzu on the epoch unbonding record
-				_, success := k.RecordsKeeper.AddHostZoneToEpochUnbondingRecord(ctx, epochUnbondingRecordId, hostZone.ChainId, hostZoneUnbonding)
-				if !success {
-					errMsg := fmt.Sprintf("Could not add host zone to epoch unbonding record | %s", err.Error())
-					k.Logger(ctx).Error(errMsg)
-					continue
-				}
-			}
+			k.RecordsKeeper.SetHostZoneUnbondings(ctx, hostZone, epochUnbondingRecordIds, recordstypes.HostZoneUnbonding_UNBONDING_IN_PROGRESS)
```
Comment (sampocs) [inline x/stakeibc/keeper/unbonding_records.go:234]: so much cleaner with a function!

File: x/records/keeper/epoch_unbonding_record.go [hunk 12]
```diff
@@ -114,3 +117,24 @@ func (k Keeper) AddHostZoneToEpochUnbondingRecord(ctx sdk.Context, epochNumber u
...
+		if !found {
+			errMsg := fmt.Sprintf("Error fetching host zone unbonding record for epoch: %d, host zone: %s", epochUnbondingRecordId, zone.ChainId)
+			k.Logger(ctx).Error(errMsg)
+		}
+		hostZoneUnbonding.Status = status
+		// save the updated hzu on the epoch unbonding record
+		_, success := k.AddHostZoneToEpochUnbondingRecord(ctx, epochUnbondingRecordId, zone.ChainId, hostZoneUnbonding)
+		if !success {
+			k.Logger(ctx).Error("Could not add host zone to epoch unbonding record")
+			continue
```
Comment (sampocs) [inline x/records/keeper/epoch_unbonding_record.go:137]: nit: rm continue since this is the last line of the loop

File: x/records/keeper/epoch_unbonding_record.go [hunk 13]
```diff
@@ -114,3 +117,24 @@ func (k Keeper) AddHostZoneToEpochUnbondingRecord(ctx sdk.Context, epochNumber u
 	}
 	return &epochUnbondingRecord, true
 }
+
+// TODO: unittest
+// TODO: what should this return?
+func (k Keeper) SetHostZoneUnbondings(ctx sdk.Context, zone stakeibctypes.HostZone, epochUnbondingRecordIds []uint64, status types.HostZoneUnbonding_Status) {
```
Comment (sampocs) [inline x/records/keeper/epoch_unbonding_record.go:123]: What is the most likely case of failure here? Seems it'd only fail if something was very off (i.e. it was passed an invalid epoch unbonding record id) - in which case we'd probably want to exit immediately, rather than continue looping through. 

If you agree with that, this should probably return an error which is checked in the calling function. And then the calling function can propagate the error if it's dealing with only one host zone, or just `continue` if it's looping host zones.

If you think there's a common error case here that doesn't necessarily indicate that something is off, and we just want to skip over it - then I think it's fine to just return nothing and take no extra action in the calling function.

File: utils/utils.go (hunk 6 above)
Comment (sampocs) [inline utils/utils.go:19]: revert

**Review by sampocs — COMMENTED** [review]
Looks great! LOVING these new status names! Also thanks for adding new unit tests!

Testing plan in the PR description looks great - have you walked through it yet or is that still a TODO?

**Comment by sampocs** [comment]
Just tested this out locally with a liquid stake + redemption flow. The deposit records status' looked great. But the unbonding record got stuck in `EXIT_TRANSFER_QUEUE` even though the unbonded tokens were successfully swept to the redemption account

File: scripts/start_network.sh (hunk 5 above)
Comment (asalzmann) [reply scripts/start_network.sh:16]: thanks done

File: x/records/keeper/callback_transfer.go (hunk 10 above)
Comment (asalzmann) [reply x/records/keeper/callback_transfer.go]: I tried that, unfortunately it does not work (see the google link above)

File: x/stakeibc/keeper/deposit_records.go (hunk 9 above)
Comment (asalzmann) [reply x/stakeibc/keeper/deposit_records.go:51]: I was thinking about this, felt slightly safer to add the check in case we change the above code. Might not be needed though, wdyt?

File: x/records/keeper/epoch_unbonding_record.go (hunk 13 above)
Comment (asalzmann) [reply x/records/keeper/epoch_unbonding_record.go:123]: Think that probably makes sense, updated to return an error

File: x/stakeibc/keeper/deposit_records.go (hunk 9 above)
Comment (sampocs) [reply x/stakeibc/keeper/deposit_records.go:51]: imo, we should just remove it. Considering it's a few lines above, I don't think there's a need to code that defensively. 

If it was hidden in another file, then I could see the case for the additional safety check

File: x/records/keeper/callback_transfer.go (hunk 10 above)
Comment (sampocs) [reply x/records/keeper/callback_transfer.go]: shucks

Comment (sampocs) [reply x/records/keeper/callback_transfer.go]: Looks like you can actually do this! Just tested it locally and it worked
```go
if _, ok := ack.Response.(* channeltypes .Acknowledgement_Error); ok {
    ...
}
```

**Comment by asalzmann** [comment]
Fixed that error, integration tests are working. Simulating the various edge cases is an outstanding todo
```
gaia_tests.bats
 ✓ [INTEGRATION-BASIC] address names are correct
 ✓ [INTEGRATION-BASIC] host zones successfully registered
 ✓ [INTEGRATION-BASIC-GAIA] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC-GAIA] liquid stake mints stATOM
 ✓ [INTEGRATION-BASIC-GAIA] tokens were transferred to GAIA after liquid staking
 ✓ [INTEGRATION-BASIC-GAIA] tokens on GAIA were staked
 ✓ [INTEGRATION-BASIC-GAIA] redemption works
 ✓ [INTEGRATION-BASIC-GAIA] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC-GAIA] rewards are being reinvested, exchange rate updating
 ✓ [NOT-IMPLEMENTED] reinvestment and revenue amounts are correct

10 tests, 0 failures
```

File: utils/utils.go (hunk 6 above)
Comment (asalzmann) [inline utils/utils.go:19]: I think the sed command might be doing this? I haven't changed this manually

Comment (asalzmann) [reply utils/utils.go:19]: @sampocs have you seen this behavior using dockernet before? My workflow is
```
make start-docker && make test-integration-docker
make stop-docker
```

Comment (sampocs) [reply utils/utils.go:19]: Now that I think about it - I think it's possible for the new address to persist if one of the build fails (because it exits before it can revert)

Comment (sampocs) [reply utils/utils.go:19]: Will send you the fix offline

File: x/records/keeper/callback_transfer.go (hunk 10 above)
Comment (sampocs) [reply x/records/keeper/callback_transfer.go]: @asalzmann bump on this

**Comment by asalzmann** [comment]
@sampocs tested all of the cases + addressed your comment

**Review by sampocs — APPROVED** [review]
